### PR TITLE
perf: improve performance of facet providers

### PR DESCRIPTION
**Pull request by pollend**
here are some minor improvements to improve cave generation. A Hashset does not seem ideal in this case especially when we can take advantage of locality. Math.hypot seems really expensive what I was looking at and reduced some of GC for that hashset that was used.

### CODE REVIEW

File: src/main/java/org/terasology/caves/CaveFacetProvider.java [hunk 1]
```diff
@@ -41,15 +42,16 @@ public void process(GeneratingRegion region) {
...
 
         // get noise in batch for performance reasons.  Getting it by individual position takes 10 times as long
-        float[][] caveNoiseValues = new float[][]{caveNoise[0].noise(facet.getWorldRegion()),caveNoise[1].noise(facet.getWorldRegion())};
+        float[][] caveNoiseValues = new float[][]{caveNoise[0].noise(facet.getWorldRegion()), caveNoise[1].noise(facet.getWorldRegion())};
 
         for (Vector3ic pos : facet.getWorldRegion()) {
             float depth = elevationFacet.getWorld(pos.x(), pos.z()) - pos.y();
             float frequencyReduction = (float) Math.max(0, 0.3 - Math.max(depth, 0) / 400); //0: no reduction, 0.7: pretty much no caves. Also somewhat increases the tendency of caves to loop rather than continuing indefinitely.
             int i = facet.getWorldIndex(pos);
-            float noiseValue = (float) Math.hypot(caveNoiseValues[0][i], caveNoiseValues[1][i]+frequencyReduction);
```
Comment (pollend) [inline src/main/java/org/terasology/caves/CaveFacetProvider.java:50]: this is significantly slower.

File: src/main/java/org/terasology/caves/CaveToSurfaceProvider.java [hunk 2]
```diff
@@ -47,7 +48,7 @@ public void process(GeneratingRegion region) {
         SurfacesFacet surfacesFacet = region.getRegionFacet(SurfacesFacet.class);
         SeaLevelFacet seaLevel = region.getRegionFacet(SeaLevelFacet.class);
 
-        Set<Vector3i> cavePositions = new HashSet<Vector3i>();
+        Set<Vector3ic> cavePositions = Sets.newHashSetWithExpectedSize(caveFacet.getWorldRegion().getSizeX() * caveFacet.getWorldRegion().getSizeZ());
```
Comment (4Denthusiast) [inline src/main/java/org/terasology/caves/CaveToSurfaceProvider.java]: Why this formula for the expected size? Some fraction of the volume would make more sense.

Comment (pollend) [reply src/main/java/org/terasology/caves/CaveToSurfaceProvider.java]: Its probably oversized. umm what fraction of the volume would be best?

Comment (skaldarnar) [reply src/main/java/org/terasology/caves/CaveToSurfaceProvider.java]: I agree that using a fraction of the region's volume would express the intent better here.

Just a wild guess: I guess we would not expect more than 50% to be cave positions? Maybe even less, so maybe 1/3? The set will grow if we add more positions, so we can play around with the value to fine tune that. 

```suggestion
        Set<Vector3ic> cavePositions = Sets.newHashSetWithExpectedSize(0.33f * caveFacet.getWorldRegion().volume());
```

File: src/main/java/org/terasology/caves/CaveToSurfaceProvider.java [hunk 3]
```diff
@@ -56,17 +57,16 @@ public void process(GeneratingRegion region) {
...
                 if (cavePositions.contains(pos)) {
                     cavePositions.remove(pos);
                     caveFacet.setWorld(pos, false);
                 }
-                if (pos.y <= seaLevel.getSeaLevel() + 1) {
+                if (pos.y() <= seaLevel.getSeaLevel() + 1) {
                     for (int x = -1; x <= 1; x++) {
                         for (int z = -1; z <= 1; z++) {
-                            Vector3i belowPos = new Vector3i(pos.x + x, pos.y - 1, pos.z + z);
+                            Vector3i belowPos = new Vector3i(pos.x() + x, pos.y() - 1, pos.z() + z);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/caves/CaveToSurfaceProvider.java]: you could reuse a `tmp` vector here if I see it correctly

**Review by skaldarnar — APPROVED** [review]

**Pull request merged**